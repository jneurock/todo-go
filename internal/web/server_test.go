package web

import (
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strconv"
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/jneurock/todo-go/internal/store"
)

func newTestServer(t *testing.T, isStoreAvailable bool) *Server {
	store := store.NewTodoInMemoryStore(isStoreAvailable)
	server, err := NewServer(store, "views")

	if err != nil {
		t.Fatal(err)
	}

	return server
}

func testHandler(t *testing.T, r *http.Request, handler func(http.ResponseWriter, *http.Request)) {
	w := httptest.NewRecorder()

	handler(w, r)

	response := w.Result()

	defer response.Body.Close()

	b, err := io.ReadAll(response.Body)

	if err != nil {
		t.Fatal(err)
	}

	snaps.MatchSnapshot(t, string(b))
}

func Test500(t *testing.T) {
	server := newTestServer(t, false)
	r := httptest.NewRequest(http.MethodGet, "/", nil)

	testHandler(t, r, server.createHandler(server.handleIndex))
}

func TestNotFound(t *testing.T) {
	server := newTestServer(t, true)
	r := httptest.NewRequest(http.MethodGet, "/doesnotexist", nil)

	testHandler(t, r, server.createHandler(server.handleIndex))
}

func TestIndex(t *testing.T) {
	server := newTestServer(t, true)

	r := httptest.NewRequest(http.MethodGet, "/", nil)

	testHandler(t, r, server.createHandler(server.handleIndex))
}

func TestIndexSeeded(t *testing.T) {
	server := newTestServer(t, true)

	server.store.Create("Do chores")
	server.store.Create("Buy groceries")

	r := httptest.NewRequest(http.MethodGet, "/", nil)

	testHandler(t, r, server.createHandler(server.handleIndex))
}

func TestNewTodo(t *testing.T) {
	server := newTestServer(t, true)
	body := strings.NewReader("description=Do%20chores")

	r := httptest.NewRequest(http.MethodPost, "/todo", body)
	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	testHandler(t, r, server.createHandler(server.handleNewTodo))
}

func TestDeleteTodo(t *testing.T) {
	server := newTestServer(t, true)
	err := server.store.Create("Delete me")

	if err != nil {
		t.Fatal(err)
	}

	todos, err := server.store.FindAll()

	if err != nil {
		t.Fatal(err)
	}

	id := strconv.FormatInt(todos[0].ID, 10)
	url := fmt.Sprintf("/todo/%s", id)

	r := httptest.NewRequest(http.MethodDelete, url, nil)
	r.SetPathValue("id", id)

	testHandler(t, r, server.createHandler(server.handleDeleteTodo))
}

func TestUpdateTodo(t *testing.T) {
	server := newTestServer(t, true)
	err := server.store.Create("Update me")

	if err != nil {
		t.Fatal(err)
	}

	todos, err := server.store.FindAll()

	if err != nil {
		t.Fatal(err)
	}

	id := strconv.FormatInt(todos[0].ID, 10)
	url := fmt.Sprintf("/todo/%s", id)
	body := strings.NewReader("complete=on&description=Update%20me%20again")

	r := httptest.NewRequest(http.MethodPut, url, body)
	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	r.SetPathValue("id", id)

	testHandler(t, r, server.createHandler(server.handleUpdateTodo))
}
