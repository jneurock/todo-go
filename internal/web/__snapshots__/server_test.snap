
[Test500 - 1]
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Todos | Something Went Wrong</title>
    <meta charset="utf-8" />
    <meta name="description" content="Something went wrong and the app is not working right now.">
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" href="/static/css/main.css" />
  </head>
  <body>
    <main class="container">
      <h1>Something Went Wrong</h1>
      <p>
        The app isn&rsquo;t working right now. Please try again later.
      </p>
    </main>
  </body>
</html>

---

[TestNotFound - 1]
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Todos | Page Not Found</title>
    <meta charset="utf-8" />
    <meta name="description" content="The page at this URL was not found. Visit the home page instead.">
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" href="/static/css/main.css" />
  </head>
  <body>
    <main class="container">
      <h1>Page Not Found</h1>
      <p>
        That&rsquo;s to be expected, honestly.
      </p>
      <p>
        This little TODO app only has one page: <a href="/">The home page</a>.
      </p>
    </main>
  </body>
</html>

---

[TestIndex - 1]
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Todos</title>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" href="/static/css/main.css" />
    <script src="/static/js/htmx.min.js"></script>
  </head>
  <body>
    <main class="container">
      <h1>Todos</h1>
      
  <div id="todos">
    <form
      class="todo-form"
      hx-post="/todo"
      hx-swap="outerHTML"
      hx-target="#todos"
    >
      <div class="todo-form--label">
        <label for="description">Description</label>
      </div>
      <div class="todo-form--controls">
        <div>
          <input
            class="todo-form--input"
            name="description"
            type="text"
            autofocus
            required
          >
          
        </div>
        <div>
          <button class="todo-form--button" type="submit">Create</button>
        </div>
      </div>
    </form>
    
      <p>Add some todos!</p>
    
  </div>

    </main>
  </body>
</html>

---

[TestIndexSeeded - 1]
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Todos</title>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" href="/static/css/main.css" />
    <script src="/static/js/htmx.min.js"></script>
  </head>
  <body>
    <main class="container">
      <h1>Todos</h1>
      
  <div id="todos">
    <form
      class="todo-form"
      hx-post="/todo"
      hx-swap="outerHTML"
      hx-target="#todos"
    >
      <div class="todo-form--label">
        <label for="description">Description</label>
      </div>
      <div class="todo-form--controls">
        <div>
          <input
            class="todo-form--input"
            name="description"
            type="text"
            autofocus
            required
          >
          
        </div>
        <div>
          <button class="todo-form--button" type="submit">Create</button>
        </div>
      </div>
    </form>
    
      
  <div class="todo-container">
    <form
      class="todo"
      hx-put="/todo/1"
      hx-swap="outerHTML"
      hx-target="#todos"
      hx-trigger="change throttle:250"
    >
      <div class="todo--attrs">
        <div>
          <input
            name="complete"
            type="checkbox"
            
          />
        </div>
        <div class="todo--description ">
          <input
            name="description"
            type="text"
            value="Buy groceries"
            required
          >
        </div>
      </div>
      <div class="todo--controls">
        <button hx-delete="/todo/1" hx-target="#todos">
          &times;
        </button>
      </div>
    </form>
    
  </div>

    
      
  <div class="todo-container">
    <form
      class="todo"
      hx-put="/todo/0"
      hx-swap="outerHTML"
      hx-target="#todos"
      hx-trigger="change throttle:250"
    >
      <div class="todo--attrs">
        <div>
          <input
            name="complete"
            type="checkbox"
            
          />
        </div>
        <div class="todo--description ">
          <input
            name="description"
            type="text"
            value="Do chores"
            required
          >
        </div>
      </div>
      <div class="todo--controls">
        <button hx-delete="/todo/0" hx-target="#todos">
          &times;
        </button>
      </div>
    </form>
    
  </div>

    
  </div>

    </main>
  </body>
</html>

---

[TestNewTodo - 1]

  <div id="todos">
    <form
      class="todo-form"
      hx-post="/todo"
      hx-swap="outerHTML"
      hx-target="#todos"
    >
      <div class="todo-form--label">
        <label for="description">Description</label>
      </div>
      <div class="todo-form--controls">
        <div>
          <input
            class="todo-form--input"
            name="description"
            type="text"
            autofocus
            required
          >
          
        </div>
        <div>
          <button class="todo-form--button" type="submit">Create</button>
        </div>
      </div>
    </form>
    
      
  <div class="todo-container">
    <form
      class="todo"
      hx-put="/todo/0"
      hx-swap="outerHTML"
      hx-target="#todos"
      hx-trigger="change throttle:250"
    >
      <div class="todo--attrs">
        <div>
          <input
            name="complete"
            type="checkbox"
            
          />
        </div>
        <div class="todo--description ">
          <input
            name="description"
            type="text"
            value="Do chores"
            required
          >
        </div>
      </div>
      <div class="todo--controls">
        <button hx-delete="/todo/0" hx-target="#todos">
          &times;
        </button>
      </div>
    </form>
    
  </div>

    
  </div>

---

[TestDeleteTodo - 1]

  <div id="todos">
    <form
      class="todo-form"
      hx-post="/todo"
      hx-swap="outerHTML"
      hx-target="#todos"
    >
      <div class="todo-form--label">
        <label for="description">Description</label>
      </div>
      <div class="todo-form--controls">
        <div>
          <input
            class="todo-form--input"
            name="description"
            type="text"
            autofocus
            required
          >
          
        </div>
        <div>
          <button class="todo-form--button" type="submit">Create</button>
        </div>
      </div>
    </form>
    
      <p>Add some todos!</p>
    
  </div>

---

[TestUpdateTodo - 1]

  <div id="todos">
    <form
      class="todo-form"
      hx-post="/todo"
      hx-swap="outerHTML"
      hx-target="#todos"
    >
      <div class="todo-form--label">
        <label for="description">Description</label>
      </div>
      <div class="todo-form--controls">
        <div>
          <input
            class="todo-form--input"
            name="description"
            type="text"
            autofocus
            required
          >
          
        </div>
        <div>
          <button class="todo-form--button" type="submit">Create</button>
        </div>
      </div>
    </form>
    
      
  <div class="todo-container">
    <form
      class="todo"
      hx-put="/todo/0"
      hx-swap="outerHTML"
      hx-target="#todos"
      hx-trigger="change throttle:250"
    >
      <div class="todo--attrs">
        <div>
          <input
            name="complete"
            type="checkbox"
            
          />
        </div>
        <div class="todo--description ">
          <input
            name="description"
            type="text"
            value="Do not update me"
            required
          >
        </div>
      </div>
      <div class="todo--controls">
        <button hx-delete="/todo/0" hx-target="#todos">
          &times;
        </button>
      </div>
    </form>
    
  </div>

    
      
  <div class="todo-container">
    <form
      class="todo"
      hx-put="/todo/1"
      hx-swap="outerHTML"
      hx-target="#todos"
      hx-trigger="change throttle:250"
    >
      <div class="todo--attrs">
        <div>
          <input
            name="complete"
            type="checkbox"
            checked
          />
        </div>
        <div class="todo--description todo--description__complete">
          <input
            name="description"
            type="text"
            value="Update me again"
            required
          >
        </div>
      </div>
      <div class="todo--controls">
        <button hx-delete="/todo/1" hx-target="#todos">
          &times;
        </button>
      </div>
    </form>
    
  </div>

    
  </div>

---
